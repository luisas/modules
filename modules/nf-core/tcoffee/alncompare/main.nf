process TCOFFEE_ALNCOMPARE {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'oras://community.wave.seqera.io/library/t-coffee_pigz:91ac7e26b23bb246':
        'community.wave.seqera.io/library/t-coffee_pigz:7d1373a24f76afe6' }"

    input:
    tuple val(meta), path(msa), path(ref_msa)

    output:
    tuple val(meta), path("*.scores"), emit: scores
    path "versions.yml"              , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ? task.ext.prefix : "${msa.baseName}"
    def args = task.ext.args ? task.ext.args.contains('compare_mode') ? task.ext.args  : (task.ext.args +'-compare_mode tc ' ) : '-compare_mode tc '
    def metric_name = args.split('compare_mode ')[1].split(' ')[0]
    def header = meta.keySet().join(",")
    def values = meta.values().join(",")

    """
    # check whether it is compressed
    if [[ "${msa}" == *.gz ]]; then
        unpigz -c ${msa} > uncompressed_msa.fa
    else
        ln ${msa} uncompressed_msa.fa
    fi

    export TEMP='./'
    t_coffee -other_pg aln_compare \
        -al1 ${ref_msa} \
        -al2 uncompressed_msa.fa \
        ${args} \
        | grep -v "seq1" | grep -v '*' | \
        awk '{ print \$4}' ORS="\t" \
        >> "scores.txt"

    # Add metadata info to output file
    echo "${header},${metric_name}" > "${prefix}.scores"

    # Add values
    scores=\$(awk '{sub(/[[:space:]]+\$/, "")} 1' scores.txt | tr -s '[:blank:]' ',')
    echo "${values},\$scores" >> "${prefix}.scores"

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        tcoffee: \$( t_coffee -version | awk '{gsub("Version_", ""); print \$3}')
        pigz: \$(echo \$(pigz --version 2>&1) | sed 's/^.*pigz\\w*//' ))
    END_VERSIONS
    """
    stub:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    # Otherwise, tcoffee will crash when calling its version
    export TEMP='./'
    touch "${prefix}.scores"

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        tcoffee: \$( t_coffee -version | awk '{gsub("Version_", ""); print \$3}')
        pigz: \$(echo \$(pigz --version 2>&1) | sed 's/^.*pigz\\w*//' ))
    END_VERSIONS
    """
}
