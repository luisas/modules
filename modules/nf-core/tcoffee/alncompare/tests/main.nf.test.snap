{
    "seatoxin - compressed": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "setoxin-ref.fa.scores:md5,c77aceec520beb56f08c342e01c56a14"
                    ]
                ],
                "1": [
                    "versions.yml:md5,bf6ee077f62052ec8e3ee8d836f50c87"
                ],
                "scores": [
                    [
                        {
                            "id": "test"
                        },
                        "setoxin-ref.fa.scores:md5,c77aceec520beb56f08c342e01c56a14"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,bf6ee077f62052ec8e3ee8d836f50c87"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-02-19T08:10:13.98747038"
    },
    "seatoxin - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.scores:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,bf6ee077f62052ec8e3ee8d836f50c87"
                ],
                "scores": [
                    [
                        {
                            "id": "test"
                        },
                        "test.scores:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,bf6ee077f62052ec8e3ee8d836f50c87"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-02-19T08:10:31.783898713"
    },
    "seatoxin": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "setoxin-ref.scores:md5,c77aceec520beb56f08c342e01c56a14"
                    ]
                ],
                "1": [
                    "versions.yml:md5,bf6ee077f62052ec8e3ee8d836f50c87"
                ],
                "scores": [
                    [
                        {
                            "id": "test"
                        },
                        "setoxin-ref.scores:md5,c77aceec520beb56f08c342e01c56a14"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,bf6ee077f62052ec8e3ee8d836f50c87"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-02-19T08:10:01.028826144"
    }
}